File: python_leaoai/repositories/Implementations/Factory/ProjectBlockFactory.py
# repositories/project_block_factory.py
from sqlalchemy.orm import Session
from typing import List, Optional
from models.factory import ProjectBlock

class ProjectBlockFactory:
    def __init__(self, db: Session):
        self.db = db

    # --- CRUD ---
    def get_obj(self, id: int) -> Optional[ProjectBlock]:
        return self.db.query(ProjectBlock).filter(ProjectBlock.id == id).first()

    def save_obj(self, obj: ProjectBlock) -> ProjectBlock:
        self.db.add(obj)
        self.db.commit()
        self.db.refresh(obj)
        return obj

    def update_obj(self, obj: ProjectBlock) -> ProjectBlock:
        self.db.merge(obj)
        self.db.commit()
        self.db.refresh(obj)
        return obj

    def get_all(self) -> List[ProjectBlock]:
        return self.db.query(ProjectBlock).all()

    def get_list(self) -> List[ProjectBlock]:
        return self.db.query(ProjectBlock).filter(ProjectBlock.excluido == False).all()

    def delete_obj(self, id: int) -> bool:
        obj = self.get_obj(id)
        if not obj:
            return False
        try:
            obj.ativo = False
            obj.excluido = True
            self.db.commit()
            return True
        except:
            self.db.rollback()
            return False

    # --- Custom Queries ---
    def get_blocks_by_project(self, project_id: int) -> List[ProjectBlock]:
        return self.db.query(ProjectBlock)\
                      .filter(ProjectBlock.project_id == project_id, ProjectBlock.excluido == False)\
                      .all()

    def get_blocks_by_project_phase(self, project_phase_id: int) -> List[ProjectBlock]:
        return self.db.query(ProjectBlock)\
                      .filter(ProjectBlock.project_phase_id == project_phase_id, ProjectBlock.excluido == False)\
                      .all()

    def get_blocks_for_conclude(self) -> List[ProjectBlock]:
        return self.db.query(ProjectBlock)\
                      .filter(ProjectBlock.block_for_conclude == True, ProjectBlock.excluido == False)\
                      .all()

    def get_block_by_code(self, code: str) -> Optional[ProjectBlock]:
        return self.db.query(ProjectBlock)\
                      .filter(ProjectBlock.code == code, ProjectBlock.excluido == False)\
                      .first()

    def get_blocks_by_user(self, usuario_id: int) -> List[ProjectBlock]:
        return self.db.query(ProjectBlock)\
                      .filter(ProjectBlock.usuario_id == usuario_id, ProjectBlock.excluido == False)\
                      .all()

File: Framework.Repositories/Implementations/Factory/ProjectBlockFactory.cs
using System.Collections.Generic;
using System.Linq;
using Framework.Data.Models.Factory;

namespace Data.Repositories
{
	public class ProjectBlockFactory
	{
		private readonly ApplicationDbContext _context;

		public ProjectBlockFactory(ApplicationDbContext context)
		{
			_context = context;
		}

		#region Crud
		public ProjectBlock GetObj(int id)
		{
			return _context.ProjectBlocks.FirstOrDefault(pb => pb.Id == id);
		}

		public ProjectBlock SaveObj(ProjectBlock obj)
		{
			_context.ProjectBlocks.Add(obj);
			_context.SaveChanges();
			return obj;
		}

		public ProjectBlock UpdateObj(ProjectBlock obj)
		{
			_context.ProjectBlocks.Update(obj);
			_context.SaveChanges();
			return obj;
		}

		public List<ProjectBlock> GetAll()
		{
			return _context.ProjectBlocks.ToList();
		}

		public List<ProjectBlock> GetList()
		{
			return _context.ProjectBlocks.Where(pb => !pb.Excluido).ToList();
		}

		public bool DeleteObj(int id)
		{
			var obj = GetObj(id);
			if (obj == null)
				return false;
			try
			{
				obj.Ativo = false;
				obj.Excluido = true;
				_context.ProjectBlocks.Update(obj);
				_context.SaveChanges();
				return true;
			}
			catch (Exception)
			{
				return false;
			}
		}
		#endregion

		#region Custom Queries
		public List<ProjectBlock> GetBlocksByProject(int projectId)
		{
			return _context.ProjectBlocks
						   .Where(pb => pb.ProjectId == projectId && !pb.Excluido)
						   .ToList();
		}

		public List<ProjectBlock> GetBlocksByProjectPhase(int projectPhaseId)
		{
			return _context.ProjectBlocks
						   .Where(pb => pb.ProjectPhaseId == projectPhaseId && !pb.Excluido)
						   .ToList();
		}

		public List<ProjectBlock> GetBlocksForConclude()
		{
			return _context.ProjectBlocks
						   .Where(pb => pb.BlockForConclude && !pb.Excluido)
						   .ToList();
		}

		public ProjectBlock GetBlockByCode(string code)
		{
			return _context.ProjectBlocks
						   .FirstOrDefault(pb => pb.Code == code && !pb.Excluido);
		}

		public List<ProjectBlock> GetBlocksByUser(int usuarioId)
		{
			return _context.ProjectBlocks
						   .Where(pb => pb.UsuarioId == usuarioId && !pb.Excluido)
						   .ToList();
		}
		#endregion
	}
}
